src/new-components/tabs/tabs.tsx

```diff
@@ -0,0 +1,197 @@
+import * as React from 'react'
+import classNames from 'classnames'
+import {
+    useTabState,
+    Tab as BaseTab,
+    TabList as BaseTabList,
+    TabPanel as BaseTabPanel,
+    TabStateReturn,
+} from 'reakit/Tab'
+import { Inline } from '../inline'
+import { usePrevious } from '../../hooks/use-previous'
+import type { ResponsiveProp } from '../responsive-props'
+import type { Space } from '../common-types'
+
+import styles from './tabs.module.css'
+
+const TabsContext = React.createContext<
+    (TabStateReturn & Omit<TabsProps, 'children' | 'selectedId'>) | null
+>(null)
+
+type TabsProps = {
+    /** The `<Tabs>` component must be composed from a `<TabList>` and corresponding `<TabPanel>` components */
+    children: React.ReactNode
+    /**
+     * Determines the primary colour of the tabs
+     */
+    color?: 'primary' | 'secondary' | 'tertiary'
+    /**
+     * Determines the style of the tabs
+     */
+    variant?: 'normal' | 'plain'
+    /**
+     * The id of the selected tab
+     */
+    selectedId?: string | null
+}
+
+/**
+ * Used to group components that compose a set of tabs. There can only be one active tab within the same `<Tabs>` group.
+ */
+function Tabs({
+    children,
+    selectedId,
+    color = 'primary',
+    variant = 'normal',
+}: TabsProps): React.ReactElement {
+    const tabState = useTabState({ selectedId })
```

Comment:
By fully owning the state management inside this component, we lose some control. I'm thinking, for instance, in the members & groups UI in the New Twist settings. In it, there was content outside the tab panels that was affected by what was the currently selected tab.

https://user-images.githubusercontent.com/15199/130153170-f0e983c1-5a93-4721-9f20-e09736a8f5be.mp4

See how, under this design, there's some UI in the unused area to the right of the tab list. This part of the UI is not rendered inside the tab panels, but the fact that I have access to the state (namely the currently active tab) I was able to conditionally render what's in that area.

I realize this is more convenient, but if a component like this is going to power all our tab-like UIs, it may need to think about this in its design, at least in the long run. It's ok if this gets this PR out of scope for your immediate needs, though.

Reply:
Yes, that's a valid use case 👍 I've added a `<TabAwareSlot>` component to provide a render prop for this

Reply:
Wow, that's a really cool solution. Thanks!